src/hotspot/share/cds/heapShared.cpp

```diff
@@ -339,6 +340,29 @@ class KlassToOopHandleTable: public ResourceHashtable<Klass*, OopHandle,
   }
 };
 
+void HeapShared::init_scratch_references() {
+  if (_scratch_references_table == nullptr)
+    _scratch_references_table = new (mtClass)ResolvedReferenceScratchTable();
```

Comment:
These two lines are outside of a lock so you could run into a race condition. I think you can remove this function and move these two lines to just before calling `_scratch_references_table->put()` in `add_scratch_resolved_references`.